### PR TITLE
Eliminate `CustomSerialize` and subsequent changes

### PR DESCRIPTION
**Pull request by MathieuDutSik**
## Motivation

The CusomSerialize is a bad design. Instead, the right way is to implement Serialize for Custom types.

## Proposal

The following is done:
* Elimination of `CustomMapView` and similar.
* Elimination of `CustomSerialize`.
* Introduction of specific `Serialize` trait for `Key` in social.
* For the matching Engine, the `PriceBid/PriceAsk` is introduced. This leads to a cleaner code. Also, the avoidance of the inversion will lead to easier-to-read GraphQL output.

## Test Plan

Unfortunately, we could not test the `matching_engine` which is currently marked as ignore and both the new and old test cannot be tested (connection lost). Test have been added for the serialize/deserialize of the `PriceAsk/PriceBid`.

## Release Plan

If the CustomSerialize is present in the documentation, then it has to be eliminated.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: examples/social/src/lib.rs [hunk 1]
```diff
@@ -251,33 +250,56 @@ pub struct Key {
...
+impl Serialize for Key {
+    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
         let data = (
             (!self.timestamp.micros()).to_be_bytes(),
             &self.author,
             (!self.index).to_be_bytes(),
         );
-        Ok(bcs::to_bytes(&data)?)
+        let key = bcs::to_bytes(&data).unwrap();
+        serializer.serialize_bytes(&key)
```
Comment (afck) [inline examples/social/src/lib.rs:261]: Aren't we already using BCS anyway?
So it shouldn't be necessary to explicitly use it here:
```suggestion
        data.serialize(serializer)
```

File: examples/social/src/lib.rs [hunk 2]
```diff
@@ -251,33 +250,56 @@ pub struct Key {
...
+            }
+
+            fn visit_bytes<E: serde::de::Error>(self, v: &[u8]) -> Result<Key, E> {
+                let (time_bytes, author, idx_bytes) =
+                    (bcs::from_bytes(v).map_err(|e| E::custom(format!("{:?}", e))))?;
+                Ok(Key {
+                    timestamp: Timestamp::from(!u64::from_be_bytes(time_bytes)),
+                    author,
+                    index: !u64::from_be_bytes(idx_bytes),
+                })
```
Comment (afck) [inline examples/social/src/lib.rs:281]: Couldn't this just be the body of `deserialize` itself? Then we wouldn't need a `Visitor` impl at all.

```suggestion
                let (time_bytes, author, idx_bytes) =
                    ([u8; 8], ChainId, [u8; 8])::deserialize(deserializer)?;
                Ok(Key {
                    timestamp: Timestamp::from(!u64::from_be_bytes(time_bytes)),
                    author,
                    index: !u64::from_be_bytes(idx_bytes),
                })
```

File: examples/matching-engine/src/lib.rs [hunk 3]
```diff
@@ -50,35 +68,80 @@ impl ServiceAbi for MatchingEngineAbi {
...
-/// We use the custom serialization for the Price so that the order of the serialization
-/// corresponds to the order of the Prices.
-impl CustomSerialize for Price {
-    fn to_custom_bytes(&self) -> Result<Vec<u8>, ViewError> {
-        let mut short_key = bcs::to_bytes(&self.price)?;
+impl Price {
+    pub fn to_bid(&self) -> PriceBid {
+        PriceBid { price: self.price }
+    }
+    pub fn to_ask(&self) -> PriceAsk {
```
Comment (afck) [inline examples/matching-engine/src/lib.rs:82]:
```suggestion

    pub fn to_ask(&self) -> PriceAsk {
```

File: examples/matching-engine/src/lib.rs [hunk 4]
```diff
@@ -50,35 +68,80 @@ impl ServiceAbi for MatchingEngineAbi {
...
+}
+
+/// We use the custom serialization for the Price so that the smallest ask price shows
+/// up first in the lexicographic order.
+impl Serialize for PriceAsk {
+    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        let mut short_key = bcs::to_bytes(&self.price).unwrap();
         short_key.reverse();
-        Ok(short_key)
+        serializer.serialize_bytes(&short_key)
```
Comment (afck) [inline examples/matching-engine/src/lib.rs:105]:
```suggestion
        self.price.to_be_bytes().serialize(serializer)
```

File: examples/matching-engine/src/lib.rs [hunk 5]
```diff
@@ -50,35 +68,80 @@ impl ServiceAbi for MatchingEngineAbi {
...
+    }
+}
+
+impl<'de1> Deserialize<'de1> for PriceAsk {
+    fn deserialize<D: serde::Deserializer<'de1>>(deserializer: D) -> Result<Self, D::Error> {
+        let result = deserializer.deserialize_byte_buf(VisitorPrice);
+        result.map(|price| Self { price })
```
Comment (afck) [inline examples/matching-engine/src/lib.rs:112]:
```suggestion
        Ok(Self { price: u64::from_be_bytes(_::deserialize(deserializer)?) })
```

File: examples/matching-engine/src/lib.rs [hunk 6]
```diff
@@ -50,35 +68,80 @@ impl ServiceAbi for MatchingEngineAbi {
...
-            price: u64::MAX - self.price,
-        }
+/// We use the custom serialization for the Price so that the highest bid price shows
+/// up first in the lexicographic order.
+impl Serialize for PriceBid {
+    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        let price = u64::MAX - self.price;
+        let mut short_key = bcs::to_bytes(&price).unwrap();
+        short_key.reverse();
+        serializer.serialize_bytes(&short_key)
```
Comment (afck) [inline examples/matching-engine/src/lib.rs:135]:
```suggestion
        (!self.price).to_be_bytes().serialize(serializer)
```

File: examples/matching-engine/src/lib.rs [hunk 7]
```diff
@@ -50,35 +68,80 @@ impl ServiceAbi for MatchingEngineAbi {
...
+    }
+}
+
+impl<'de1> Deserialize<'de1> for PriceBid {
+    fn deserialize<D: serde::Deserializer<'de1>>(deserializer: D) -> Result<Self, D::Error> {
+        let result = deserializer.deserialize_byte_buf(VisitorPrice);
+        result.map(|price| Self {
+            price: u64::MAX - price,
+        })
```
Comment (afck) [inline examples/matching-engine/src/lib.rs:144]:
```suggestion
        Ok(Self { price: !u64::from_be_bytes(_::deserialize(deserializer)?) })
```

**Review by ma2bd — CHANGES_REQUESTED** [review]
I agree that we could remove `CustomSerialize` but in the current state of `MapView` the solution here only applies to fixed size objects. (Otherwise, after serializing values to a BCS vector,  the BCS-encoded lengths, serialized first, will ruin ordering.)

Incidentally, for many fixed size objects such as integers, it would be slightly better to shuffle the bytes but still serialize as an integer.

Following @Twey's idea, I suspect a better approach might be to provide only `MapView` but (optionally or always?) use an order-preserving encoding instead of BCS.

**Comment by MathieuDutSik** [comment]
That PR is closed because it combines several unrelated issues.
Also, the issues of serialization are not well though out.

**Pull request closed without merging**